HIP/hip-1028.md

```diff
@@ -0,0 +1,157 @@
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2024-08-15
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/pull/1028
+requires: 646, 657, 765
+updated: 2024-08-19
+requested-by: NA
+---
+
+## Abstract
+
+HIP-646/657/765 introduces a new metadata field for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key  
+for updating metadata. However, these features are not supported by smart contracts.  
+This proposal aims to enhance Hedera Smart Contract Service (HSCS) by exposing HIP-646/657/765 features.
+
+## Motivation
+
+A user may need to create, update and get info for Fungible Tokens (FT), Non-Fungible Tokens (NFT) and METADATA key through contract.  
+The current version of Hedera Smart Contract Service (HSCS) does not cover these use cases.
+
+## Rationale
+
+The HSCS already has the functionality to create and update tokens, so we would want to extend these.  
+
+However, we will add new functions for getting info for tokens to not break the older version of ``HederaTokenService.sol``  
+
+New function  for TokenUpdateNfts HAPI operation will be added  
+
+## User stories
+
+1. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with metadata.
+2. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata.
+3. As an EOA I would like to initiate a smart contract transaction that updates a fungible token with metadata.
+4. As an EOA I would like to initiate a smart contract transaction that updates a non-fungible token collection with metadata.
+5. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with a metadata key.
+6. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata key.
+7. As an EOA I would like to initiate a smart contract transaction that updates a fungible token with metadata key.
+8. As an EOA I would like to initiate a smart contract transaction that updates a non-fungible token collection with metadata key.
+9. As an EOA I would like to initiate a smart contract transaction that creates a fungible token with metadata and metadata key.
+10. As an EOA I would like to initiate a smart contract transaction that creates a non-fungible token collection with metadata and metadata key.
+11. As an EOA I would like to initiate a smart contract transaction that updates the metadata of NFT(s).
+
+## Specification
+
+We will utilize the existing Hedera Token Service (HTS) HAPI transactions.  
+To achieve this the Hedera Token Service (HTS) system contract must be updated to expose the necessary features.  
+Existing system contract functions must be extended and new functions must be added to the `IHederaTokenService` interface  
```

Comment:
Above mention is made of `HederaTokenService.sol`, here `IHederaTokenService.sol`.  Perhaps here should _also_ be mentioned `HederaTokenService.sol`?  (Maybe as a footnote too a reminder to the reader of what each one is _for_.)